redskyctl/internal/commands/experiments/suggest.go

```diff
@@ -181,66 +177,71 @@ func (o *SuggestOptions) assignInteractive(p *experimentsv1alpha1.Parameter, def
 		}
 		text := s.Text()
 		if text == "" && def != nil {
-			return *def, nil
+			return def, nil
 		}
-		number, err := checkValue(p, json.Number(text))
+		result, err := checkValue(p, text)
 		if err != nil {
 			continue
 		}
-		return number, nil
+		return result, nil
 	}
 
 	if err := s.Err(); err != nil {
-		return "0", err
+		return nil, err
 	}
-	return "0", fmt.Errorf("no assignment for parameter: %s", p.Name)
+	return nil, fmt.Errorf("no assignment for parameter: %s", p.Name)
 }
 
-func checkValue(p *experimentsv1alpha1.Parameter, n json.Number) (json.Number, error) {
-	switch p.Type {
-	case experimentsv1alpha1.ParameterTypeInteger:
-		min, max, err := intBounds(&p.Bounds)
-		if err != nil {
-			return "0", err
-		}
-		v, err := n.Int64()
-		if err != nil {
-			return "0", err
-		}
-		if v < min || v > max {
-			return "0", fmt.Errorf("value is not within experiment bounds [%d-%d]: %d", min, max, v)
-		}
-	case experimentsv1alpha1.ParameterTypeDouble:
-		min, max, err := floatBounds(&p.Bounds)
-		if err != nil {
-			return "0.0", err
-		}
-		v, err := n.Float64()
-		if err != nil {
-			return "0.0", err
-		}
-		if v < min || v > max {
-			return "0.0", fmt.Errorf("value is not within experiment bounds [%f-%f]: %f", min, max, v)
-		}
+func prompt(p *experimentsv1alpha1.Parameter, def *numstr.NumberOrString) string {
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("Assignment for %v parameter '%s'", p.Type, p.Name))
+
+	// Add the bounds
+	if p.Type == experimentsv1alpha1.ParameterTypeCategorical {
+		b.WriteString(fmt.Sprintf(" [%s]", strings.Join(p.Values, ", ")))
+	} else if p.Bounds != nil {
+		b.WriteString(fmt.Sprintf(" [%v,%v]", p.Bounds.Min, p.Bounds.Max))
+	}
+
+	// Add the default
+	if def != nil {
+		b.WriteString(fmt.Sprintf(" (%s)", def.String()))
+	}
+
+	b.WriteString(": ")
+	return b.String()
+}
+
+func checkValue(p *experimentsv1alpha1.Parameter, s string) (*numstr.NumberOrString, error) {
+	v, err := p.ParseValue(s)
+	if err != nil {
+		return nil, err
 	}
-	return n, nil
+	err = experimentsv1alpha1.CheckParameterValue(p, v)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
 }
 
-func randomValue(p *experimentsv1alpha1.Parameter) (*json.Number, error) {
+func randomValue(p *experimentsv1alpha1.Parameter) (*numstr.NumberOrString, error) {
 	switch p.Type {
 	case experimentsv1alpha1.ParameterTypeInteger:
-		min, max, err := intBounds(&p.Bounds)
+		min, max, err := intBounds(p.Bounds)
 		if err != nil {
 			return nil, err
 		}
-		r := json.Number(strconv.FormatInt(rand.Int63n(max-min)+min, 10))
+		r := numstr.FromInt64(rand.Int63n(max-min) + min)
```

Comment:
not related to this pr, but do we seed the rand here?

Reply:
We seed the global rand cryptographically in the `init()` of the main package.